eng/Migrate-Package.ps1

```diff
@@ -0,0 +1,204 @@
+)
+
+#### Functions ####
+
+Function Write-Color
+{
+    Param (
+        [ValidateNotNullOrWhiteSpace()]
+        [string] $newColor
+    )
+
+    $oldColor = $host.UI.RawUI.ForegroundColor
+    $host.UI.RawUI.ForegroundColor = $newColor
+
+    If ($args)
+    {
+        Write-Output $args
+    }
+    Else
+    {
+        $input | Write-Output
+    }
+
+    $host.UI.RawUI.ForegroundColor = $oldColor
+}
+
+Function VerifyPathOrExit
+{
+    Param (
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $path
+    )
+
+    If (-Not (Test-Path -Path $path))
+    {
+        Write-Error "The path does not exist: $path"
+        Exit
+    }
+    ElseIf (-Not ($path -match "^[a-zA-Z0-9\.\-_:/ ]+$"))
+    {
+        Write-Error "The path will not work with Git. Avoid backslashes: $path"
+        Exit
+    }
+
+    Write-Color green "The path is valid: $path"
+}
+
+Function Execute-Command
+{
+    Param (
+        [Parameter(Mandatory=$true)]
+        [ValidateNotNullOrWhiteSpace()]
+        [string]
+        $command
+    )
+
+    Write-Color cyan "Executing: $command"
+    Invoke-Expression $command
+}
+
+#### Execution ####
+
+Write-Output "Verifying paths..."
+
+$OriginRepoPath = [IO.Path]::TrimEndingDirectorySeparator($OriginRepoPath)
+$DestinationRepoPath = [IO.Path]::TrimEndingDirectorySeparator($DestinationRepoPath)
+$AssemblyFileOrDirectoryRelativeDirectoryPath = [IO.Path]::TrimEndingDirectorySeparator($AssemblyFileOrDirectoryRelativeDirectoryPath)
+$FullAssemblyPath = "$OriginRepoPath/$AssemblyFileOrDirectoryRelativeDirectoryPath/$AssemblyFileOrDirectoryToMigrate"
+$RelativeAssemblyPath = "$AssemblyFileOrDirectoryRelativeDirectoryPath/$AssemblyFileOrDirectoryToMigrate"
+
+VerifyPathOrExit $OriginRepoPath
+VerifyPathOrExit $DestinationRepoPath
+VerifyPathOrExit $FullAssemblyPath
+
+$SplittedBranchName = "SplittedBranch"
+$MigratedFolderName = "migrated"
+$WorkingBranchName = "WorkingBranch"
+$NoHistoryBranchName = "NoHistoryBranch"
+$OldRepoRemoteName = "old-repo"
+$NewRepoRemoteName = "new-repo"
+
+# Go to origin repo
+Execute-Command "Set-Location $OriginRepoPath -Verbose"
+
+# Get the branch just in case
+Execute-Command "git fetch $OriginRemote $OriginBranch"
+
+# Switch to the desired branch
+Execute-Command "git checkout $OriginBranch"
+
+# Delete the existing branch, just in case
+Execute-Command "git branch -D $SplittedBranchName"
+
+# Filter commits relevant to assembly, may take a long time
+Execute-Command "git subtree split -P '$RelativeAssemblyPath' -b $SplittedBranchName"
+
+# Move to that branch
+Execute-Command "git checkout $SplittedBranchName"
+
+# Remove any files that don't belong
+Execute-Command "git reset --hard"
+
+# Put everything under the migrated directory
+Execute-Command "New-Item -ItemType Directory -Name '$MigratedFolderName' -Verbose"
+Execute-Command "Move-Item -Path '*' -Destination '$MigratedFolderName' -Exclude '$MigratedFolderName' -Verbose"
```

Comment:
I need to double check if this command needs the `-Recurse` argument. I just saw an error when testing it, but it might have been because I forgot to clean the destination folder.